### PR TITLE
Add actions for testing

### PR DESCRIPTION
**Pull request by neriyaco**

### CODE REVIEW

**Comment by elazarcoh** [comment]
The action would not work out of the box. We need to update it depending on how we would build the package. I'm currently working on a cross platform, portable way to do it, using mesonpy. This PR should wait for it, I think.

**Comment by elazarcoh** [comment]
Also, need to check how to test it on multiple platforms (win, linux, osx)

**Comment by neriyaco** [comment]
Ok I understand, I will put this PR on hold for now.


> Also, need to check how to test it on multiple platforms (win, linux, osx)

GitHub Actions has support for all operating systems, after we decide on the build process I'll create a proper workflow for testing.

**Comment by elazarcoh** [comment]
See #15

File: .github/workflows/python-test.yml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+on:
+  push:
+    branches: [ "main" ]
+  pull_request:
+    branches: [ "main" ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
```
Comment (elazarcoh) [inline .github/workflows/python-test.yml:15]:
```suggestion
    runs-on: ${{ matrix.os }}
```

File: .github/workflows/python-test.yml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version: ["3.7", "3.8", "3.9", "3.10"]
```
Comment (elazarcoh) [inline .github/workflows/python-test.yml:19]:
```suggestion
        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11"]
```

File: .github/workflows/python-test.yml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  pull_request:
+    branches: [ "main" ]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
```
Comment (elazarcoh) [inline .github/workflows/python-test.yml:18]:
```suggestion
      matrix:
          os: [ubuntu-latest, macos-latest, windows-latest]
```

File: .github/workflows/python-test.yml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    steps:
+    - uses: actions/checkout@v3
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v3
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        python -m pip install flake8 pytest hatch
```
Comment (elazarcoh) [inline .github/workflows/python-test.yml:30]:
```suggestion
        python -m pip install flake8 pytest build
```

File: .github/workflows/python-test.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Lint with flake8
+      run: |
+        # stop the build if there are Python syntax errors or undefined names
+        flake8 . --count --select=E9,F63,F7,F82 --show-source --statistics
+        # exit-zero treats all errors as warnings. The GitHub editor is 127 chars wide
+        flake8 . --count --exit-zero --max-complexity=10 --max-line-length=127 --statistics
+    - name: Build package
+      run: |
+      python -m hatch run build
```
Comment (elazarcoh) [inline .github/workflows/python-test.yml:40]:
```suggestion
      python -m build
```

**Comment by elazarcoh** [comment]
@neriyaco rebase on `main` and "force push (with lease)

**Comment by neriyaco** [comment]
Looks like building on linux/mac does not work with meson.

**Comment by neriyaco** [comment]
I did some tests and comparison between my local machine and GitHub Actions, the difference is the compiler, my local machine uses MSVC and GitHub uses gcc for Windows and Linux and clang for MacOS.
I'll let @binyamin555 handle this as he's the one writing the C++ code.

**Comment by neriyaco** [comment]
Now the Windows build uses MSVC, and it looks like it's working.
I assume most Windows developers use MSVC so that's the main reason for that.

**Comment by elazarcoh** [comment]
There are a few gcc errors (and warning).
I'll try to build in linux and fix it.

**Review by binyamin555 — APPROVED** [review]

**Review by elazarcoh — APPROVED** [review]

File: requirements.txt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+attrs==22.2.0
```
[the remark this answers is not in the document]
Comment (neriyaco) [reply requirements.txt:1]: I'm not sure if it's mandatory, we need to tell python to install typing-extensions that's all

**Pull request merged**